sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs

```diff
@@ -88,5 +85,158 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+
+            public async ValueTask<(string? headerValue, bool scheduleRefreshToken)> GetHeaderValueAsync(bool async, CancellationToken cancellationToken)
+            {
+                (string? headerValue, Task<string>? pendingTask, bool scheduleRefreshToken) = GetHeaderValue();
+                if (pendingTask == null)
+                {
+                    return (headerValue, scheduleRefreshToken);
+                }
+
+                if (async)
+                {
+                    return (await pendingTask.AwaitWithCancellation(cancellationToken), false);
+                }
+
+                try
+                {
+                    pendingTask.Wait(cancellationToken);
+                }
+                catch (AggregateException) { } // ignore exception here to rethrow it with EnsureCompleted
+
+                return (pendingTask.EnsureCompleted(), false);
+            }
+
+            public string SaveToken(AccessToken token)
+            {
+                if (string.IsNullOrEmpty(token.Token))
+                {
+                    var exception = new InvalidOperationException($"{nameof(TokenCredential)}.{nameof(TokenCredential.GetToken)} has failed with unknown error.");
+                    ResetTokenIfExpired(exception);
+                    throw exception;
+                }
+
+                string headerValue;
+                TaskCompletionSource<string>? pendingTcs;
+
+                lock (_syncObj)
+                {
+                    _headerValue = "Bearer " + token.Token;
+                    _refreshOn = token.ExpiresOn - _tokenRefreshOffset;
+                    _expiresOn = token.ExpiresOn;
+                    _tokenState = TokenState.Valid;
+
+                     pendingTcs = _pendingTcs;
+                    headerValue = _headerValue;
+                    _pendingTcs = null;
+                }
+
+                pendingTcs?.SetResult(headerValue);
+                return headerValue;
+            }
+
+            public void ResetTokenIfExpired(Exception exception)
+            {
+                TaskCompletionSource<string>? pendingTcs = default;
+
+                lock (_syncObj)
+                {
+                    if (DateTimeOffset.UtcNow >= _expiresOn)
+                    {
+                        _headerValue = default;
+                        _refreshOn = default;
+                        _expiresOn = default;
+                        _tokenState = TokenState.Invalid;
+
+                        pendingTcs = _pendingTcs;
+                        _pendingTcs = null;
+                    }
+                    else if (_tokenState == TokenState.AboutToExpire)
+                    {
+                        _tokenState = TokenState.Valid;
+                        if (_pendingTcs != default)
+                        {
+                            exception = new InvalidOperationException($"{nameof(GetHeaderValueAsync)} shouldn't wait in {TokenState.AboutToExpire} state.", exception);
```

Comment:
Should you do this check and throw before you reset the state to valid? Otherwise is the cache in an inconsistent state if the exception is handled?

Reply:
No, but I need to set `_pendingTcs` to null.  We should get out of the `TokenState.AboutToExpire` state anyway.

Reply:
Fixed